File: commons-cli-graal/src/main/java/org/apache/commons/cli/OptionGroup.java
/*
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

package org.apache.commons.cli;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.graalvm.polyglot.PolyglotException;
import org.graalvm.polyglot.Value;

/**
 * A group of mutually exclusive options.
 */
public class OptionGroup implements Serializable {
    /** The serial version UID. */
    private static final long serialVersionUID = 1L;

    /** handle to the Python class */
    private static Value clz = ContextInitializer.getPythonClass("option_group.py", "OptionGroup");

    /** handle to the Python object */
    private Value obj;

    /**
     * Create an option group.
     */
    public OptionGroup() {
        obj = clz.execute();
    }

    /**
     * (Re)create an option group with the python object.
     *
     * @param foreignOptionGroup the option to add to this group
     */
    public OptionGroup(final Value foreignOptionGroup) {
        obj = foreignOptionGroup;
    }

    /** Cache for wrappers to python objects  */
    private static Map<Value, OptionGroup> cache = new HashMap<>();

    public static OptionGroup create(final Value foreignOptionGroup) {
        if (foreignOptionGroup.isNull()) {
            return null;
        }
        return cache.computeIfAbsent(foreignOptionGroup, OptionGroup::new);
    }

    public Value getPythonObject() {
        return obj;
    }

    /**
     * Add the specified {@code Option} to this group.
     *
     * @param option the option to add to this group
     * @return this option group with the option added
     */
    public OptionGroup addOption(final Option option) {
        obj.invokeMember("add_option", option);
        return this;
    }

    /**
     * @return the names of the options in this group as a {@code Collection}
     */
    public Collection<String> getNames() {
        // the key set is the collection of names
        return obj.invokeMember("get_names").as(List.class);
    }

    /**
     * @return the options in this group as a {@code Collection}
     */
    public Collection<Option> getOptions() {
        // the values are the collection of python objects
        return IntegrationUtil.valueArrayToCollection(obj.invokeMember("get_options"), Option.class, ArrayList.class);
    }

    /**
     * @return the selected option name
     */
    public String getSelected() {
        return obj.getMember("selected").asString();
    }

    /**
     * Tests whether this option group is required.
     *
     * @return whether this option group is required
     */
    public boolean isRequired() {
        return obj.getMember("required").asBoolean();
    }

    /**
     * @param required specifies if this group is required
     */
    public void setRequired(final boolean required) {
        obj.invokeMember("set_required", required);
    }

    /**
     * Set the selected option of this group to {@code name}.
     *
     * @param option the option that is selected
     * @throws AlreadySelectedException if an option from this group has already been selected.
     */
    public void setSelected(final Option option) throws AlreadySelectedException {
        try {
            obj.invokeMember("set_selected", option);
        } catch (PolyglotException e) {
            throw new AlreadySelectedException(this, option);
        }
    }

    /**
     * Returns the stringified version of this OptionGroup.
     *
     * @return the stringified representation of this group
     */
    @Override
    public String toString() {
        return obj.invokeMember("to_string").asString();
    }
}

File: commons-cli-python/src/main/python/option_group.py
from main.python.java_handler import java_handler
from main.python.option import Option

@java_handler
class OptionGroup:
    """
    A group of mutually exclusive options.
    """
    def __init__(self):
        self.serial_version_uid = 1
        self.option_Map = {}
        self.required = False
    
    def add_option(self, option: Option):
        """
        Add the specified option to this group.
        :param option: the option to add to this group
        :return: this option group with the option added
        """
        # key - option name
        # value - the option
        self.option_Map[option.get_key()] = option
        return self

    def get_names(self):
        """
        :return: the names of the options in this group as a collection
        """
        # the key set is the collection of names
        return list(self.option_Map.keys())

    def get_options(self):
        """
        :return: the options in this group as a collection
        """
        # the values are the collection of options
        return list(self.option_Map.values())
    
    def get_selected(self):
        """
        :return: the selected option name
        """
        return self.selected
    
    def is_required(self):
        """
        Tests whether this option group is required.
        :return: whether this option group is required
        """
        return self.required
    
    def set_required(self, required: bool):
        """
        :param required: specifies if this group is required
        """
        self.required = required
    
    def set_selected(self, option: Option):
        """
        Set the selected option of this group to option.
        :param option: the option that is selected
        :raises AlreadySelectedException: if an option from this group has already been selected.
        """
        if option is None:
            # reset the option previously selected
            self.selected = None
            return
        
        # if no option has already been selected or the
        # same option is being reselected then set the
        # selected member variable
        from main.python.already_selected_exception import AlreadySelectedException
        
        option_key = option.get_key()

        if self.selected is not None and self.selected != option_key:
            raise AlreadySelectedException(self, option)

        self.selected = option_key

    def __str__(self):
        """
        Returns the stringified version of this OptionGroup.
        :return: the stringified representation of this group
        """
        buff = "["
        for i, option in enumerate(self.get_options()):
            if option.get_opt():
                buff += "-"
                buff += option.get_opt()
            else:
                buff += "--"
                buff += option.get_long_opt()

            if option.get_description():
                buff += " "
                buff += option.get_description()

            if i < len(self.get_options()) - 1:
                buff += ", "

        buff += "]"

        return buff

    def to_string(self):
        return str(self)
